atom/browser/lib/power_observer.h

```diff
@@ -18,7 +18,19 @@ namespace atom {
 #if defined(OS_LINUX)
 typedef PowerObserverLinux PowerObserver;
 #else
-typedef base::PowerObserver PowerObserver;
+class PowerObserver : public base::PowerObserver {
+ public:
+  PowerObserver() {}
+  void BlockShutdown() {}
+  void UnblockShutdown() {}
+  // Notification that the system is rebooting or shutting down. If the
+  // implementation returns true, the PowerObserver instance should try to delay
+  // OS shutdown so the application can perform cleanup before exiting.
+  virtual bool OnShutdown() { return false; }
```

Comment:
This code only works if the caller and the called are the same object (an instance of PowerMonitor inheriting from PowerObserver).. It's kind of a [template method](https://en.wikipedia.org/wiki/Template_method_pattern) via a crazy dependency between observer and monitor. Likewise, this is why Observer has new verb methods in this PR.

Mind you, I understand the choice -- we've already got the DBus code in Observer, so unless we refactor the DBus code into another place (eg a self-contained class or in PowerMonitorLinux) then Monitor has to have these Observer hooks.

This is going to be a pain to make future changes in and I'd love it if we could somehow clean up this dependency chain in this PR. That's a big ask though.

Reply:
I'm not sure I understand your request as I'm not very skilled in C++ design patterns. Can you give an example of what you have in mind?